### PR TITLE
Feat/implement chat logic

### LINKED ISSUE #199: fullstack: Implement chat logic
### **Description**:  
Once **issue #198** is completed (Chat UI), implement the **full chat functionality** using **Firebase Realtime** database features. The chat interface must support sending and receiving messages in real time and use the existing internal contact list to initiate new conversations. 

The implementation must be **highly efficient, responsive, and user-friendly**, with smooth message flow and updates.

---

### **Requirements**:  
1. **Wait for UI Completion**:  
   - This issue must begin **only after issue #198 is finished and merged**.

2. **Firebase Integration**:  
   - Use **Firebase Realtime updates** (obligatory).
   - Configure Firestore listeners to fetch chat threads and new messages in real time.

3. **Firestore Structure**:
   - **Chat Collection**: 
   ```typescript
     /chats/{chatId}
     /chats/{chatId}/messages/{messageId}
   ```
     - Each chat must contain:  
     ```typescript
     {
       "participants": ["user1_uid", "user2_uid"],
       "lastMessage": "Hola!",
       "updatedAt": timestamp
     }
     ```
   - **Messages Subcollection**: under each chat
     - Each message must contain:  
     ```typescript
     {
     "senderId": "user1_uid",
     "text": "Hola!",
     "createdAt": timestamp
     }
      ```

4. **Sending Messages**:
   - Allow users to send new messages within a chat thread.
   - New messages must be added in Firestore under the correct `messages` subcollection.
   - Update `lastMessage` and `updatedAt` in the parent chat document.

5. **Starting New Conversations**:
   - User selects from **existing internal contacts** (fetched via:) EXAMPLE IN: 
   `src\components\modules\escrow\hooks\multi-release\initialize-multi-escrow.hook.ts`
   ```typescript
     const fetchContacts = useContactStore((state) => state.fetchContacts);
     const contacts = useContactStore((state) => state.contacts);
   ```
   - Once a contact is selected:
     - If a chat with that user exists → open it.
     - If not → create a new chat document with both users as participants.

6. **UX/UI Behavior**:
   - Auto-scroll on new messages.
   - Display message timestamps.
   - Show read/delivered status (if available in schema).
   - Handle disconnections or permission errors gracefully.

7. **Code Structure**:
   - Follow the module pattern from other parts of the app.
   - Implement chat logic inside the `/modules/chat` folder.
   - Use hooks or helpers to manage listeners and Firebase writes.

---

### **Validations**:  
1. Real-time messages are received instantly.
2. Messages are stored in the correct chat document structure.
3. Starting a chat via contacts works correctly.
4. Sending a message updates the last message and timestamp in the chat thread.
5. Error handling and edge cases (e.g., no internet, permission denied) are managed properly.

---

### **Tests**:  
1. **Realtime Messaging**: Open two sessions and verify message sync between them.
2. **New Chat Flow**: Select contact, create chat, send and receive messages.
3. **Message Storage**: Verify Firestore writes are correct and efficient.
4. **UI Sync**: Confirm chat UI updates properly when messages are sent/received.
5. **Fallbacks**: Check behavior on Firebase failure or denied permissions.

⚠️ **Please follow the guidelines for requesting an issue, launching commits, and describing PRs. Otherwise, they will be ignored.** ⚠️

### PR DESCRIPTION
**Pull request by JosueBrenes**
<p align="center"> 
  CLR-S (2) 
</p>

# Pull Request | Trustless Work

## 1. Issue Link

- Closes #199 

---

## 2. Brief Description of the Issue

This pull request introduces realtime chat functionality using Firebase Realtime Database and Firestore. It enables message sending with file attachments (of any type), visual display of attachments (images or download links), and full conversation deletion support.

---

## 3. Type of Change

Mark with an `x` all checkboxes that apply.

- [ ] 📝 Documentation (updates to README, docs, or comments)  
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)  
- [x] 👌 Enhancement (non-breaking change which adds functionality)  
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)  

---

## 4. Changes Made

- Extended the `Message` type to include optional attachment metadata.  
- Added client-side support for reading and sending files with messages.  
- Chat messages display images inline or offer download links for attachments.  
- "Delete Conversation" option added to chat header dropdown.  
- Firestore service includes real-time listeners and a `deleteChat` helper function.  

---

## 5. Evidence After Solution

https://github.com/user-attachments/assets/6fbff282-2fe2-4e7c-9832-a0a5592209b4

---

## Summary by CodeRabbit

* **New Features**
  * Added the ability to start new chat conversations from a contact list.
  * Enabled sending messages with file attachments, including inline image previews and downloadable links for other file types.
  * Introduced the option to delete chat conversations directly from the chat interface.

* **Enhancements**
  * Chat messages and conversations now update in real-time with Firebase integration.
  * Improved file attachment validation to avoid erroneous rejections when no file type restrictions are set.

* **User Interface**
  * Updated chat input and message display to support attachments.
  * Added a "New Chat" button and popover for starting conversations.
  * Added delete conversation option to chat header dropdown menu.

### CODE REVIEW

File: src/components/modules/chat/hooks/chat.hook.ts [hunk 1]
```diff
@@ -1,22 +1,83 @@
 "use client";
 
-import { useState, useCallback } from "react";
+import { useState, useCallback, useEffect } from "react";
 import type { Conversation, ChatMessage } from "@/@types/conversation.entity";
+import { useContactStore } from "@/core/store/data/slices/contacts.slice";
+import { useGlobalAuthenticationStore } from "@/core/store/data";
 import {
-  MOCK_CONVERSATIONS,
-  MOCK_CHAT_MESSAGES,
```
Comment (JoelVR17) [inline src/components/modules/chat/hooks/chat.hook.ts:7]: Please delete the mock data files.

File: src/components/modules/chat/server/chat.firebase.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  query,
+  serverTimestamp,
+  updateDoc,
+  where,
+  DocumentData,
+} from "firebase/firestore";
+
+const chatsCollection = collection(db, "chats");
+
+export interface Chat {
```
Comment (JoelVR17) [inline src/components/modules/chat/server/chat.firebase.ts:19]: Move these entities to the correct file 

`src/@types/conversation.entity.ts`

**Review by JoelVR17 — CHANGES_REQUESTED** [review]
Please resolve these issues. Your work was great, thanks!
